{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "work_dir = '/home/jovyan/work/'\n",
    "data_dir = path.join(work_dir,'processed_data')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join(work_dir,'external')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata.tsv'),index_col=0,sep='\\t')\n",
    "\n",
    "for idx, row in df_metadata[df_metadata['treatment'].isnull()].iterrows():\n",
    "    df_metadata.loc[idx, 'treatment'] = row.condition\n",
    "df_metadata.drop('condition', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine different values to create unique condition names\n",
    "cat_cols = ['treatment', 'growth_phase', 'isolate', 'genetic_changes'] \n",
    "\n",
    "df_metadata['condition'] = df_metadata['base_media'].str.cat(df_metadata[cat_cols],sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CM_1</th>\n",
       "      <td>CcpA CodY interaction</td>\n",
       "      <td>CA-MHB;Control;exp;JE2;WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM_10</th>\n",
       "      <td>CcpA CodY interaction</td>\n",
       "      <td>CA-MHB;None;exp;JE2;marR_TF::bursa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM_11</th>\n",
       "      <td>CcpA CodY interaction</td>\n",
       "      <td>CA-MHB;None;exp;JE2;ilvA::bursa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM_12</th>\n",
       "      <td>CcpA CodY interaction</td>\n",
       "      <td>CA-MHB;None;exp;JE2;ilvA::bursa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM_13</th>\n",
       "      <td>CcpA CodY interaction</td>\n",
       "      <td>CA-MHB;2g/Lglucose;exp;JE2;ilvA::bursa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     project                               condition\n",
       "CM_1   CcpA CodY interaction               CA-MHB;Control;exp;JE2;WT\n",
       "CM_10  CcpA CodY interaction      CA-MHB;None;exp;JE2;marR_TF::bursa\n",
       "CM_11  CcpA CodY interaction         CA-MHB;None;exp;JE2;ilvA::bursa\n",
       "CM_12  CcpA CodY interaction         CA-MHB;None;exp;JE2;ilvA::bursa\n",
       "CM_13  CcpA CodY interaction  CA-MHB;2g/Lglucose;exp;JE2;ilvA::bursa"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata.to_csv(path.join(data_dir,'metadata.tsv'),sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your TRN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the TRN file represents a regulatory interaction.  \n",
    "**Your TRN file must have the following columns:**\n",
    "1. `regulator` - Name of regulator (`/` or `+` characters will be converted to `;`)\n",
    "1. `gene_id` - Locus tag of gene being regulated\n",
    "\n",
    "The following columns are optional, but are helpful to have:\n",
    "1. `regulator_id` - Locus tag of regulator\n",
    "1. `gene_name` - Name of gene (can automatically update this using `name2num`)\n",
    "1. `direction` - Direction of regulation ('+' for activation, '-' for repression, '?' or NaN for unknown)\n",
    "1. `evidence` - Evidence of regulation (e.g. ChIP-exo, qRT-PCR, SELEX, Motif search)\n",
    "1. `PMID` - Reference for regulation\n",
    "\n",
    "You may add any other columns that could help you. TRNs may be saved as either CSV or TSV files. See below for an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>TF_id</th>\n",
       "      <th>trn_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>N315_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zur</td>\n",
       "      <td>SA1383</td>\n",
       "      <td>TF</td>\n",
       "      <td>SA2203</td>\n",
       "      <td>SA2203</td>\n",
       "      <td>USA300HOU_RS13055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zur</td>\n",
       "      <td>SA1383</td>\n",
       "      <td>TF</td>\n",
       "      <td>SA0806</td>\n",
       "      <td>SA0806</td>\n",
       "      <td>USA300HOU_RS04685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zur</td>\n",
       "      <td>SA1383</td>\n",
       "      <td>TF</td>\n",
       "      <td>rpsN2</td>\n",
       "      <td>SA1171</td>\n",
       "      <td>USA300HOU_RS06795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zur</td>\n",
       "      <td>SA1383</td>\n",
       "      <td>TF</td>\n",
       "      <td>rpmG2</td>\n",
       "      <td>SAS042</td>\n",
       "      <td>USA300HOU_RS06790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zur</td>\n",
       "      <td>SA1383</td>\n",
       "      <td>TF</td>\n",
       "      <td>znuC</td>\n",
       "      <td>SA1385</td>\n",
       "      <td>USA300HOU_RS08305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulator   TF_id trn_type gene_name N315_gene_id            gene_id\n",
       "0       Zur  SA1383       TF    SA2203       SA2203  USA300HOU_RS13055\n",
       "1       Zur  SA1383       TF    SA0806       SA0806  USA300HOU_RS04685\n",
       "2       Zur  SA1383       TF     rpsN2       SA1171  USA300HOU_RS06795\n",
       "3       Zur  SA1383       TF     rpmG2       SAS042  USA300HOU_RS06790\n",
       "4       Zur  SA1383       TF      znuC       SA1385  USA300HOU_RS08305"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_trn = pd.read_csv(path.join(external_data,'TRN.csv'))\n",
    "df_trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `regulator` and `gene_id` must be filled in for each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_trn.regulator.notnull().all())\n",
    "print(df_trn.gene_id.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:The characters '+' and '/' are used for combining regulons and cannot be in regulator names. These characters will be replaced with ';'\n",
      "WARNING:root:The following genes are in the TRN but not in your M matrix: {' USA300HOU_1986', ' USA300HOU_1483', ' USA300HOU_1442', 'USA300HOU_RS06665', ' USA300HOU_1973', ' USA300HOU_1952', ' USA300HOU_2009', ' USA300HOU_1435', ' USA300HOU_1967', ' USA300HOU_1990', ' USA300HOU_0858', ' USA300HOU_0860', ' USA300HOU_1476', ' USA300HOU_1968', ' USA300HOU_1432', ' USA300HOU_1440', ' USA300HOU_1487', ' USA300HOU_0852', ' USA300HOU_1463', ' USA300HOU_1988', ' USA300HOU_2004', ' USA300HOU_1992', ' USA300HOU_1457', ' USA300HOU_1950', ' USA300HOU_1484', ' USA300HOU_1477', ' USA300HOU_1987', ' USA300HOU_1976', ' USA300HOU_1467', ' USA300HOU_1473', ' USA300HOU_1453', ' USA300HOU_1494', ' USA300HOU_0848', ' USA300HOU_2006', ' USA300HOU_1959', ' USA300HOU_1495', ' USA300HOU_1438', ' USA300HOU_1444', ' USA300HOU_1954', ' USA300HOU_1497', ' USA300HOU_2013', ' USA300HOU_1447', ' USA300HOU_1437', ' USA300HOU_1479', ' USA300HOU_1989', ' USA300HOU_1469', ' USA300HOU_1947', ' USA300HOU_1982', ' USA300HOU_1430', ' USA300HOU_1946', ' USA300HOU_1468', ' USA300HOU_2005', ' USA300HOU_1981', ' USA300HOU_1944', ' USA300HOU_1470', ' USA300HOU_1949', ' USA300HOU_1972', ' USA300HOU_1486', ' USA300HOU_1985', ' USA300HOU_1445', ' USA300HOU_1481', 'USA300HOU_RS10560', ' USA300HOU_1965', ' USA300HOU_1945', ' USA300HOU_1493', ' USA300HOU_1960', ' USA300HOU_2008', ' USA300HOU_1478', ' USA300HOU_2001', ' USA300HOU_1970', ' USA300HOU_1957', ' USA300HOU_1433', ' USA300HOU_1446', ' USA300HOU_1461', 'USA300HOU_RS10920', ' USA300HOU_1456', ' USA300HOU_1431', ' USA300HOU_1963', ' USA300HOU_1956', ' USA300HOU_1488', ' USA300HOU_1460', ' USA300HOU_1458', ' USA300HOU_1975', ' USA300HOU_1466', ' USA300HOU_1464', ' USA300HOU_1995', ' USA300HOU_1964', ' USA300HOU_1998', ' USA300HOU_1489', ' USA300HOU_1465', ' USA300HOU_1471', ' USA300HOU_2003', ' USA300HOU_0868', ' USA300HOU_1439', ' USA300HOU_1459', 'USA300HOU_RS09585', ' USA300HOU_0863', 'USA300HOU_RS10540', ' USA300HOU_0855', ' USA300HOU_2012', ' USA300HOU_1448', ' USA300HOU_0859', ' USA300HOU_1999', ' USA300HOU_1955', ' USA300HOU_1966', ' USA300HOU_1980', ' USA300HOU_1978', ' USA300HOU_1490', ' USA300HOU_0861', ' USA300HOU_1993', ' USA300HOU_1951', ' USA300HOU_0867', ' USA300HOU_1441', ' USA300HOU_1969', ' USA300HOU_1475', ' USA300HOU_1971', ' USA300HOU_1942', ' USA300HOU_0857', ' USA300HOU_0850', ' USA300HOU_1997', ' USA300HOU_1452', ' USA300HOU_1498', ' USA300HOU_0853', ' USA300HOU_1434', ' USA300HOU_1462', ' USA300HOU_0854', ' USA300HOU_1962', ' USA300HOU_2002', ' USA300HOU_0862', ' USA300HOU_0849', 'USA300HOU_RS07965', ' USA300HOU_1994', ' USA300HOU_1984', ' USA300HOU_0865', 'USA300HOU_RS09610', ' USA300HOU_0869', 'USA300HOU_RS12850', ' USA300HOU_1974', ' USA300HOU_0866', ' USA300HOU_1979', ' USA300HOU_1449', ' USA300HOU_1496', ' USA300HOU_1436', 'USA300HOU_RS00225', ' USA300HOU_1482', 'USA300HOU_RS06650', ' USA300HOU_1953', ' USA300HOU_1454', ' USA300HOU_1492', ' USA300HOU_0856', ' USA300HOU_1991', ' USA300HOU_2011', ' USA300HOU_0851', ' USA300HOU_1472', ' USA300HOU_1451', ' USA300HOU_1977', ' USA300HOU_0870', ' USA300HOU_0864', ' USA300HOU_1961', ' USA300HOU_1983', ' USA300HOU_1443', ' USA300HOU_1996', ' USA300HOU_1485', ' USA300HOU_0847', ' USA300HOU_2000', ' USA300HOU_2007', ' USA300HOU_1450', ' USA300HOU_1455', ' USA300HOU_1474', ' USA300HOU_1958'}\n",
      "WARNING:root:Optimizing iModulon thresholds, may take 2-3 minutes...\n"
     ]
    }
   ],
   "source": [
    "ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "                   A = path.join(data_dir,'A.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "                   gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "                   trn = path.join(external_data,'TRN.csv'),\n",
    "                   optimize_cutoff=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulatory iModulons\n",
    "Use `compute_trn_enrichment` to automatically check for Regulatory iModulons. The more complete your TRN, the more regulatory iModulons you'll find."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>FruR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.977615e-13</td>\n",
       "      <td>3.955230e-13</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.076433</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>12.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>6.486677e-39</td>\n",
       "      <td>1.946003e-38</td>\n",
       "      <td>0.838710</td>\n",
       "      <td>0.351351</td>\n",
       "      <td>0.495238</td>\n",
       "      <td>26.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>VraR</td>\n",
       "      <td>3.907315e-06</td>\n",
       "      <td>5.860973e-06</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.164384</td>\n",
       "      <td>6.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>Purine</td>\n",
       "      <td>2.523321e-11</td>\n",
       "      <td>7.569963e-11</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9</td>\n",
       "      <td>TPP</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>PurR</td>\n",
       "      <td>1.365061e-29</td>\n",
       "      <td>2.730122e-29</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>11</td>\n",
       "      <td>VraS2</td>\n",
       "      <td>9.312770e-09</td>\n",
       "      <td>2.793831e-08</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>6.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>1.213330e-81</td>\n",
       "      <td>2.426660e-80</td>\n",
       "      <td>0.654545</td>\n",
       "      <td>0.595041</td>\n",
       "      <td>0.623377</td>\n",
       "      <td>72.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>MalR</td>\n",
       "      <td>1.341246e-13</td>\n",
       "      <td>1.341246e-12</td>\n",
       "      <td>0.081818</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.151261</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15</td>\n",
       "      <td>CggR</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>17</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>5.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.656345e-07</td>\n",
       "      <td>8.281726e-07</td>\n",
       "      <td>0.179487</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>7.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>22</td>\n",
       "      <td>FMN</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>23</td>\n",
       "      <td>ArcR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>23</td>\n",
       "      <td>ArgR</td>\n",
       "      <td>2.758501e-12</td>\n",
       "      <td>5.517002e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.312500</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>23</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.498836e-10</td>\n",
       "      <td>1.998448e-10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.151515</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>23</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>1.255425e-07</td>\n",
       "      <td>1.255425e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.079365</td>\n",
       "      <td>5.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24</td>\n",
       "      <td>PyrR</td>\n",
       "      <td>8.232349e-23</td>\n",
       "      <td>1.646470e-22</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>24</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.832865e-06</td>\n",
       "      <td>1.832865e-06</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.038217</td>\n",
       "      <td>0.072289</td>\n",
       "      <td>6.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>28</td>\n",
       "      <td>Genomic Island 3</td>\n",
       "      <td>2.803906e-19</td>\n",
       "      <td>5.607811e-19</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.540541</td>\n",
       "      <td>10.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>29</td>\n",
       "      <td>CodY</td>\n",
       "      <td>3.741945e-50</td>\n",
       "      <td>4.864528e-49</td>\n",
       "      <td>0.616279</td>\n",
       "      <td>0.360544</td>\n",
       "      <td>0.454936</td>\n",
       "      <td>53.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>29</td>\n",
       "      <td>Lysine</td>\n",
       "      <td>6.003517e-10</td>\n",
       "      <td>3.902286e-09</td>\n",
       "      <td>0.081395</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.147368</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>29</td>\n",
       "      <td>T-box(Trp)</td>\n",
       "      <td>4.265891e-09</td>\n",
       "      <td>1.848553e-08</td>\n",
       "      <td>0.069767</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>29</td>\n",
       "      <td>HisR</td>\n",
       "      <td>2.594151e-08</td>\n",
       "      <td>6.744792e-08</td>\n",
       "      <td>0.081395</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>0.141414</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>29</td>\n",
       "      <td>T-box(Met)</td>\n",
       "      <td>2.199368e-08</td>\n",
       "      <td>6.744792e-08</td>\n",
       "      <td>0.058140</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.109890</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>31</td>\n",
       "      <td>CodY</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>9.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>32</td>\n",
       "      <td>CtsR</td>\n",
       "      <td>2.899696e-26</td>\n",
       "      <td>8.699089e-26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>32</td>\n",
       "      <td>HrcA</td>\n",
       "      <td>9.534712e-13</td>\n",
       "      <td>1.430207e-12</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>37</td>\n",
       "      <td>CymR</td>\n",
       "      <td>1.655744e-29</td>\n",
       "      <td>4.967233e-29</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.717949</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>38</td>\n",
       "      <td>Genomic Island 1</td>\n",
       "      <td>1.040356e-36</td>\n",
       "      <td>3.121069e-36</td>\n",
       "      <td>0.354839</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>22.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>39</td>\n",
       "      <td>AgrA</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>2.652405e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>40</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>0.283784</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>41</td>\n",
       "      <td>HssR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>43</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.602650e-16</td>\n",
       "      <td>2.241060e-15</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>0.148649</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>46</td>\n",
       "      <td>MntR</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>48</td>\n",
       "      <td>VraS2</td>\n",
       "      <td>1.639155e-36</td>\n",
       "      <td>6.556622e-36</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.564706</td>\n",
       "      <td>24.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>48</td>\n",
       "      <td>VraR</td>\n",
       "      <td>1.401629e-24</td>\n",
       "      <td>2.803259e-24</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>18.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>49</td>\n",
       "      <td>WalR</td>\n",
       "      <td>2.244531e-30</td>\n",
       "      <td>1.571172e-29</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.178344</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>28.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>49</td>\n",
       "      <td>SaeR</td>\n",
       "      <td>6.008128e-28</td>\n",
       "      <td>2.102845e-27</td>\n",
       "      <td>0.485714</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>0.539683</td>\n",
       "      <td>17.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>50</td>\n",
       "      <td>NreC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>50</td>\n",
       "      <td>Rex</td>\n",
       "      <td>5.856495e-25</td>\n",
       "      <td>8.784742e-25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>12.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>50</td>\n",
       "      <td>VraR</td>\n",
       "      <td>6.598882e-11</td>\n",
       "      <td>6.598882e-11</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.259259</td>\n",
       "      <td>7.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>52</td>\n",
       "      <td>ManR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>54</td>\n",
       "      <td>CodY</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.168675</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>55</td>\n",
       "      <td>Fur</td>\n",
       "      <td>1.358287e-60</td>\n",
       "      <td>4.074860e-60</td>\n",
       "      <td>0.649123</td>\n",
       "      <td>0.787234</td>\n",
       "      <td>0.711538</td>\n",
       "      <td>37.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>58</td>\n",
       "      <td>LacR</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>59</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.105581e-17</td>\n",
       "      <td>5.527907e-17</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.418605</td>\n",
       "      <td>9.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>62</td>\n",
       "      <td>L20_leader</td>\n",
       "      <td>1.440816e-07</td>\n",
       "      <td>5.763265e-07</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>63</td>\n",
       "      <td>WalR</td>\n",
       "      <td>2.557458e-17</td>\n",
       "      <td>2.813204e-16</td>\n",
       "      <td>0.385714</td>\n",
       "      <td>0.171975</td>\n",
       "      <td>0.237885</td>\n",
       "      <td>27.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>65</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>4.631188e-17</td>\n",
       "      <td>3.704950e-16</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.123967</td>\n",
       "      <td>0.211268</td>\n",
       "      <td>15.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>65</td>\n",
       "      <td>MalR</td>\n",
       "      <td>1.352868e-14</td>\n",
       "      <td>5.411472e-14</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>65</td>\n",
       "      <td>TreR</td>\n",
       "      <td>3.421939e-07</td>\n",
       "      <td>9.125170e-07</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>66</td>\n",
       "      <td>Fur</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>0.290909</td>\n",
       "      <td>8.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>67</td>\n",
       "      <td>ArgR</td>\n",
       "      <td>9.774854e-09</td>\n",
       "      <td>3.909942e-08</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>68</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.012067e-10</td>\n",
       "      <td>8.096536e-10</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0.101911</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>16.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>69</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049587</td>\n",
       "      <td>0.094488</td>\n",
       "      <td>6.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>73</td>\n",
       "      <td>CodY</td>\n",
       "      <td>7.090098e-18</td>\n",
       "      <td>1.418020e-17</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.172840</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>73</td>\n",
       "      <td>Lysine</td>\n",
       "      <td>6.104774e-08</td>\n",
       "      <td>6.104774e-08</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon         regulator        pvalue        qvalue  precision  \\\n",
       "0          1              FruR  0.000000e+00  0.000000e+00   1.000000   \n",
       "1          3              WalR  1.977615e-13  3.955230e-13   0.800000   \n",
       "2          5  Genomic Island 9  6.486677e-39  1.946003e-38   0.838710   \n",
       "3          5              VraR  3.907315e-06  5.860973e-06   0.193548   \n",
       "4          6            Purine  2.523321e-11  7.569963e-11   0.500000   \n",
       "5          9               TPP  0.000000e+00  0.000000e+00   1.000000   \n",
       "6         10              PurR  1.365061e-29  2.730122e-29   0.857143   \n",
       "7         11             VraS2  9.312770e-09  2.793831e-08   0.500000   \n",
       "8         13              CcpA  1.213330e-81  2.426660e-80   0.654545   \n",
       "9         13              MalR  1.341246e-13  1.341246e-12   0.081818   \n",
       "10        15              CggR  3.789150e-15  3.789150e-15   1.000000   \n",
       "11        17              CcpA  2.460824e-06  2.460824e-06   0.714286   \n",
       "12        20               Rex  1.656345e-07  8.281726e-07   0.179487   \n",
       "13        22               FMN  3.789150e-15  3.789150e-15   0.833333   \n",
       "14        23              ArcR  0.000000e+00  0.000000e+00   1.000000   \n",
       "15        23              ArgR  2.758501e-12  5.517002e-12   1.000000   \n",
       "16        23               Rex  1.498836e-10  1.998448e-10   1.000000   \n",
       "17        23              CcpA  1.255425e-07  1.255425e-07   1.000000   \n",
       "18        24              PyrR  8.232349e-23  1.646470e-22   0.888889   \n",
       "19        24              WalR  1.832865e-06  1.832865e-06   0.666667   \n",
       "20        28  Genomic Island 3  2.803906e-19  5.607811e-19   0.625000   \n",
       "21        29              CodY  3.741945e-50  4.864528e-49   0.616279   \n",
       "22        29            Lysine  6.003517e-10  3.902286e-09   0.081395   \n",
       "23        29        T-box(Trp)  4.265891e-09  1.848553e-08   0.069767   \n",
       "24        29              HisR  2.594151e-08  6.744792e-08   0.081395   \n",
       "25        29        T-box(Met)  2.199368e-08  6.744792e-08   0.058140   \n",
       "26        31              CodY  1.987481e-12  1.987481e-12   1.000000   \n",
       "27        32              CtsR  2.899696e-26  8.699089e-26   1.000000   \n",
       "28        32              HrcA  9.534712e-13  1.430207e-12   0.500000   \n",
       "29        37              CymR  1.655744e-29  4.967233e-29   0.700000   \n",
       "30        38  Genomic Island 1  1.040356e-36  3.121069e-36   0.354839   \n",
       "31        39              AgrA  1.326202e-14  2.652405e-14   0.714286   \n",
       "32        40  Genomic Island 9  5.380175e-33  5.380175e-33   0.913043   \n",
       "33        41              HssR  0.000000e+00  0.000000e+00   1.000000   \n",
       "34        43  Genomic Island 9  5.602650e-16  2.241060e-15   0.785714   \n",
       "35        46              MntR  4.115537e-09  4.115537e-09   0.750000   \n",
       "36        48             VraS2  1.639155e-36  6.556622e-36   0.600000   \n",
       "37        48              VraR  1.401629e-24  2.803259e-24   0.450000   \n",
       "38        49              WalR  2.244531e-30  1.571172e-29   0.800000   \n",
       "39        49              SaeR  6.008128e-28  2.102845e-27   0.485714   \n",
       "40        50              NreC  0.000000e+00  0.000000e+00   1.000000   \n",
       "41        50               Rex  5.856495e-25  8.784742e-25   1.000000   \n",
       "42        50              VraR  6.598882e-11  6.598882e-11   0.583333   \n",
       "43        52              ManR  0.000000e+00  0.000000e+00   1.000000   \n",
       "44        54              CodY  4.597413e-15  4.597413e-15   0.736842   \n",
       "45        55               Fur  1.358287e-60  4.074860e-60   0.649123   \n",
       "46        58              LacR  1.326202e-14  1.326202e-14   0.714286   \n",
       "47        59               Rex  1.105581e-17  5.527907e-17   0.900000   \n",
       "48        62        L20_leader  1.440816e-07  5.763265e-07   0.187500   \n",
       "49        63              WalR  2.557458e-17  2.813204e-16   0.385714   \n",
       "50        65              CcpA  4.631188e-17  3.704950e-16   0.714286   \n",
       "51        65              MalR  1.352868e-14  5.411472e-14   0.333333   \n",
       "52        65              TreR  3.421939e-07  9.125170e-07   0.142857   \n",
       "53        66               Fur  2.876360e-15  2.876360e-15   1.000000   \n",
       "54        67              ArgR  9.774854e-09  3.909942e-08   0.666667   \n",
       "55        68              WalR  1.012067e-10  8.096536e-10   0.390244   \n",
       "56        69              CcpA  5.104426e-09  5.104426e-09   1.000000   \n",
       "57        73              CodY  7.090098e-18  1.418020e-17   0.933333   \n",
       "58        73            Lysine  6.104774e-08  6.104774e-08   0.266667   \n",
       "\n",
       "      recall   f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0   1.000000  1.000000   3.0           3.0            3.0     1.0  \n",
       "1   0.076433  0.139535  12.0         157.0           15.0     1.0  \n",
       "2   0.351351  0.495238  26.0          74.0           31.0     1.0  \n",
       "3   0.142857  0.164384   6.0          42.0           31.0     1.0  \n",
       "4   1.000000  0.666667   4.0           4.0            8.0     1.0  \n",
       "5   1.000000  1.000000   7.0           7.0            7.0     1.0  \n",
       "6   0.857143  0.857143  12.0          14.0           14.0     1.0  \n",
       "7   0.133333  0.210526   6.0          45.0           12.0     1.0  \n",
       "8   0.595041  0.623377  72.0         121.0          110.0     1.0  \n",
       "9   1.000000  0.151261   9.0           9.0          110.0     1.0  \n",
       "10  0.833333  0.909091   5.0           6.0            5.0     1.0  \n",
       "11  0.041322  0.078125   5.0         121.0            7.0     1.0  \n",
       "12  0.212121  0.194444   7.0          33.0           39.0     1.0  \n",
       "13  1.000000  0.909091   5.0           5.0            6.0     1.0  \n",
       "14  1.000000  1.000000   5.0           5.0            5.0     1.0  \n",
       "15  0.312500  0.476190   5.0          16.0            5.0     1.0  \n",
       "16  0.151515  0.263158   5.0          33.0            5.0     1.0  \n",
       "17  0.041322  0.079365   5.0         121.0            5.0     1.0  \n",
       "18  1.000000  0.941176   8.0           8.0            9.0     1.0  \n",
       "19  0.038217  0.072289   6.0         157.0            9.0     1.0  \n",
       "20  0.476190  0.540541  10.0          21.0           16.0     1.0  \n",
       "21  0.360544  0.454936  53.0         147.0           86.0     1.0  \n",
       "22  0.777778  0.147368   7.0           9.0           86.0     1.0  \n",
       "23  0.857143  0.129032   6.0           7.0           86.0     1.0  \n",
       "24  0.538462  0.141414   7.0          13.0           86.0     1.0  \n",
       "25  1.000000  0.109890   5.0           5.0           86.0     1.0  \n",
       "26  0.061224  0.115385   9.0         147.0            9.0     1.0  \n",
       "27  0.769231  0.869565  10.0          13.0           10.0     1.0  \n",
       "28  0.833333  0.625000   5.0           6.0           10.0     1.0  \n",
       "29  0.736842  0.717949  14.0          19.0           20.0     1.0  \n",
       "30  0.916667  0.511628  22.0          24.0           62.0     1.0  \n",
       "31  1.000000  0.833333   5.0           5.0            7.0     1.0  \n",
       "32  0.283784  0.432990  21.0          74.0           23.0     1.0  \n",
       "33  1.000000  1.000000   2.0           2.0            2.0     1.0  \n",
       "34  0.148649  0.250000  11.0          74.0           14.0     1.0  \n",
       "35  0.750000  0.750000   3.0           4.0            4.0     1.0  \n",
       "36  0.533333  0.564706  24.0          45.0           40.0     1.0  \n",
       "37  0.428571  0.439024  18.0          42.0           40.0     1.0  \n",
       "38  0.178344  0.291667  28.0         157.0           35.0     1.0  \n",
       "39  0.607143  0.539683  17.0          28.0           35.0     1.0  \n",
       "40  1.000000  1.000000  12.0          12.0           12.0     1.0  \n",
       "41  0.363636  0.533333  12.0          33.0           12.0     1.0  \n",
       "42  0.166667  0.259259   7.0          42.0           12.0     1.0  \n",
       "43  1.000000  1.000000   2.0           2.0            2.0     1.0  \n",
       "44  0.095238  0.168675  14.0         147.0           19.0     1.0  \n",
       "45  0.787234  0.711538  37.0          47.0           57.0     1.0  \n",
       "46  1.000000  0.833333   5.0           5.0            7.0     1.0  \n",
       "47  0.272727  0.418605   9.0          33.0           10.0     1.0  \n",
       "48  1.000000  0.315789   3.0           3.0           16.0     1.0  \n",
       "49  0.171975  0.237885  27.0         157.0           70.0     1.0  \n",
       "50  0.123967  0.211268  15.0         121.0           21.0     1.0  \n",
       "51  0.777778  0.466667   7.0           9.0           21.0     1.0  \n",
       "52  1.000000  0.250000   3.0           3.0           21.0     1.0  \n",
       "53  0.170213  0.290909   8.0          47.0            8.0     1.0  \n",
       "54  0.250000  0.363636   4.0          16.0            6.0     1.0  \n",
       "55  0.101911  0.161616  16.0         157.0           41.0     1.0  \n",
       "56  0.049587  0.094488   6.0         121.0            6.0     1.0  \n",
       "57  0.095238  0.172840  14.0         147.0           15.0     1.0  \n",
       "58  0.444444  0.333333   4.0           9.0           15.0     1.0  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also search for AND/OR combinations of regulators using the `max_regs` argument.\n",
    "\n",
    "Regulator enrichments can be directly saved to the `imodulon_table` using the `save` argument. This saves the enrichment with the lowest q-value to the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>FruR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.977615e-13</td>\n",
       "      <td>3.955230e-13</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.076433</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>12.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>6.486677e-39</td>\n",
       "      <td>1.946003e-38</td>\n",
       "      <td>0.838710</td>\n",
       "      <td>0.351351</td>\n",
       "      <td>0.495238</td>\n",
       "      <td>26.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>VraR</td>\n",
       "      <td>3.907315e-06</td>\n",
       "      <td>5.860973e-06</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.164384</td>\n",
       "      <td>6.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>Purine</td>\n",
       "      <td>2.523321e-11</td>\n",
       "      <td>7.569963e-11</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9</td>\n",
       "      <td>TPP</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>PurR</td>\n",
       "      <td>1.365061e-29</td>\n",
       "      <td>2.730122e-29</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>11</td>\n",
       "      <td>VraS2</td>\n",
       "      <td>9.312770e-09</td>\n",
       "      <td>2.793831e-08</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>6.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>1.213330e-81</td>\n",
       "      <td>2.426660e-80</td>\n",
       "      <td>0.654545</td>\n",
       "      <td>0.595041</td>\n",
       "      <td>0.623377</td>\n",
       "      <td>72.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>MalR</td>\n",
       "      <td>1.341246e-13</td>\n",
       "      <td>1.341246e-12</td>\n",
       "      <td>0.081818</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.151261</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15</td>\n",
       "      <td>CggR</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>17</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>5.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.656345e-07</td>\n",
       "      <td>8.281726e-07</td>\n",
       "      <td>0.179487</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>7.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>22</td>\n",
       "      <td>FMN</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>23</td>\n",
       "      <td>ArcR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>23</td>\n",
       "      <td>ArgR</td>\n",
       "      <td>2.758501e-12</td>\n",
       "      <td>5.517002e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.312500</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>23</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.498836e-10</td>\n",
       "      <td>1.998448e-10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.151515</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>23</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>1.255425e-07</td>\n",
       "      <td>1.255425e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.079365</td>\n",
       "      <td>5.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24</td>\n",
       "      <td>PyrR</td>\n",
       "      <td>8.232349e-23</td>\n",
       "      <td>1.646470e-22</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>24</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.832865e-06</td>\n",
       "      <td>1.832865e-06</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.038217</td>\n",
       "      <td>0.072289</td>\n",
       "      <td>6.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>28</td>\n",
       "      <td>Genomic Island 3</td>\n",
       "      <td>2.803906e-19</td>\n",
       "      <td>5.607811e-19</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.540541</td>\n",
       "      <td>10.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>29</td>\n",
       "      <td>CodY</td>\n",
       "      <td>3.741945e-50</td>\n",
       "      <td>4.864528e-49</td>\n",
       "      <td>0.616279</td>\n",
       "      <td>0.360544</td>\n",
       "      <td>0.454936</td>\n",
       "      <td>53.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>29</td>\n",
       "      <td>Lysine</td>\n",
       "      <td>6.003517e-10</td>\n",
       "      <td>3.902286e-09</td>\n",
       "      <td>0.081395</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.147368</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>29</td>\n",
       "      <td>T-box(Trp)</td>\n",
       "      <td>4.265891e-09</td>\n",
       "      <td>1.848553e-08</td>\n",
       "      <td>0.069767</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>29</td>\n",
       "      <td>HisR</td>\n",
       "      <td>2.594151e-08</td>\n",
       "      <td>6.744792e-08</td>\n",
       "      <td>0.081395</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>0.141414</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>29</td>\n",
       "      <td>T-box(Met)</td>\n",
       "      <td>2.199368e-08</td>\n",
       "      <td>6.744792e-08</td>\n",
       "      <td>0.058140</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.109890</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>31</td>\n",
       "      <td>CodY</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>9.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>32</td>\n",
       "      <td>CtsR</td>\n",
       "      <td>2.899696e-26</td>\n",
       "      <td>8.699089e-26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>32</td>\n",
       "      <td>HrcA</td>\n",
       "      <td>9.534712e-13</td>\n",
       "      <td>1.430207e-12</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>37</td>\n",
       "      <td>CymR</td>\n",
       "      <td>1.655744e-29</td>\n",
       "      <td>4.967233e-29</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.717949</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>38</td>\n",
       "      <td>Genomic Island 1</td>\n",
       "      <td>1.040356e-36</td>\n",
       "      <td>3.121069e-36</td>\n",
       "      <td>0.354839</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>22.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>39</td>\n",
       "      <td>AgrA</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>2.652405e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>40</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>0.283784</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>41</td>\n",
       "      <td>HssR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>43</td>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.602650e-16</td>\n",
       "      <td>2.241060e-15</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>0.148649</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>46</td>\n",
       "      <td>MntR</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>48</td>\n",
       "      <td>VraS2</td>\n",
       "      <td>1.639155e-36</td>\n",
       "      <td>6.556622e-36</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.564706</td>\n",
       "      <td>24.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>48</td>\n",
       "      <td>VraR</td>\n",
       "      <td>1.401629e-24</td>\n",
       "      <td>2.803259e-24</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>18.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>49</td>\n",
       "      <td>WalR</td>\n",
       "      <td>2.244531e-30</td>\n",
       "      <td>1.571172e-29</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.178344</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>28.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>49</td>\n",
       "      <td>SaeR</td>\n",
       "      <td>6.008128e-28</td>\n",
       "      <td>2.102845e-27</td>\n",
       "      <td>0.485714</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>0.539683</td>\n",
       "      <td>17.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>50</td>\n",
       "      <td>NreC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>50</td>\n",
       "      <td>Rex</td>\n",
       "      <td>5.856495e-25</td>\n",
       "      <td>8.784742e-25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>12.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>50</td>\n",
       "      <td>VraR</td>\n",
       "      <td>6.598882e-11</td>\n",
       "      <td>6.598882e-11</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.259259</td>\n",
       "      <td>7.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>52</td>\n",
       "      <td>ManR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>54</td>\n",
       "      <td>CodY</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.168675</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>55</td>\n",
       "      <td>Fur</td>\n",
       "      <td>1.358287e-60</td>\n",
       "      <td>4.074860e-60</td>\n",
       "      <td>0.649123</td>\n",
       "      <td>0.787234</td>\n",
       "      <td>0.711538</td>\n",
       "      <td>37.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>58</td>\n",
       "      <td>LacR</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>59</td>\n",
       "      <td>Rex</td>\n",
       "      <td>1.105581e-17</td>\n",
       "      <td>5.527907e-17</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.418605</td>\n",
       "      <td>9.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>62</td>\n",
       "      <td>L20_leader</td>\n",
       "      <td>1.440816e-07</td>\n",
       "      <td>5.763265e-07</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>63</td>\n",
       "      <td>WalR</td>\n",
       "      <td>2.557458e-17</td>\n",
       "      <td>2.813204e-16</td>\n",
       "      <td>0.385714</td>\n",
       "      <td>0.171975</td>\n",
       "      <td>0.237885</td>\n",
       "      <td>27.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>65</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>4.631188e-17</td>\n",
       "      <td>3.704950e-16</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.123967</td>\n",
       "      <td>0.211268</td>\n",
       "      <td>15.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>65</td>\n",
       "      <td>MalR</td>\n",
       "      <td>1.352868e-14</td>\n",
       "      <td>5.411472e-14</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>65</td>\n",
       "      <td>TreR</td>\n",
       "      <td>3.421939e-07</td>\n",
       "      <td>9.125170e-07</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>66</td>\n",
       "      <td>Fur</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>0.290909</td>\n",
       "      <td>8.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>67</td>\n",
       "      <td>ArgR</td>\n",
       "      <td>9.774854e-09</td>\n",
       "      <td>3.909942e-08</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>68</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.012067e-10</td>\n",
       "      <td>8.096536e-10</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0.101911</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>16.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>69</td>\n",
       "      <td>CcpA</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049587</td>\n",
       "      <td>0.094488</td>\n",
       "      <td>6.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>73</td>\n",
       "      <td>CodY</td>\n",
       "      <td>7.090098e-18</td>\n",
       "      <td>1.418020e-17</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.172840</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>73</td>\n",
       "      <td>Lysine</td>\n",
       "      <td>6.104774e-08</td>\n",
       "      <td>6.104774e-08</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon         regulator        pvalue        qvalue  precision  \\\n",
       "0          1              FruR  0.000000e+00  0.000000e+00   1.000000   \n",
       "1          3              WalR  1.977615e-13  3.955230e-13   0.800000   \n",
       "2          5  Genomic Island 9  6.486677e-39  1.946003e-38   0.838710   \n",
       "3          5              VraR  3.907315e-06  5.860973e-06   0.193548   \n",
       "4          6            Purine  2.523321e-11  7.569963e-11   0.500000   \n",
       "5          9               TPP  0.000000e+00  0.000000e+00   1.000000   \n",
       "6         10              PurR  1.365061e-29  2.730122e-29   0.857143   \n",
       "7         11             VraS2  9.312770e-09  2.793831e-08   0.500000   \n",
       "8         13              CcpA  1.213330e-81  2.426660e-80   0.654545   \n",
       "9         13              MalR  1.341246e-13  1.341246e-12   0.081818   \n",
       "10        15              CggR  3.789150e-15  3.789150e-15   1.000000   \n",
       "11        17              CcpA  2.460824e-06  2.460824e-06   0.714286   \n",
       "12        20               Rex  1.656345e-07  8.281726e-07   0.179487   \n",
       "13        22               FMN  3.789150e-15  3.789150e-15   0.833333   \n",
       "14        23              ArcR  0.000000e+00  0.000000e+00   1.000000   \n",
       "15        23              ArgR  2.758501e-12  5.517002e-12   1.000000   \n",
       "16        23               Rex  1.498836e-10  1.998448e-10   1.000000   \n",
       "17        23              CcpA  1.255425e-07  1.255425e-07   1.000000   \n",
       "18        24              PyrR  8.232349e-23  1.646470e-22   0.888889   \n",
       "19        24              WalR  1.832865e-06  1.832865e-06   0.666667   \n",
       "20        28  Genomic Island 3  2.803906e-19  5.607811e-19   0.625000   \n",
       "21        29              CodY  3.741945e-50  4.864528e-49   0.616279   \n",
       "22        29            Lysine  6.003517e-10  3.902286e-09   0.081395   \n",
       "23        29        T-box(Trp)  4.265891e-09  1.848553e-08   0.069767   \n",
       "24        29              HisR  2.594151e-08  6.744792e-08   0.081395   \n",
       "25        29        T-box(Met)  2.199368e-08  6.744792e-08   0.058140   \n",
       "26        31              CodY  1.987481e-12  1.987481e-12   1.000000   \n",
       "27        32              CtsR  2.899696e-26  8.699089e-26   1.000000   \n",
       "28        32              HrcA  9.534712e-13  1.430207e-12   0.500000   \n",
       "29        37              CymR  1.655744e-29  4.967233e-29   0.700000   \n",
       "30        38  Genomic Island 1  1.040356e-36  3.121069e-36   0.354839   \n",
       "31        39              AgrA  1.326202e-14  2.652405e-14   0.714286   \n",
       "32        40  Genomic Island 9  5.380175e-33  5.380175e-33   0.913043   \n",
       "33        41              HssR  0.000000e+00  0.000000e+00   1.000000   \n",
       "34        43  Genomic Island 9  5.602650e-16  2.241060e-15   0.785714   \n",
       "35        46              MntR  4.115537e-09  4.115537e-09   0.750000   \n",
       "36        48             VraS2  1.639155e-36  6.556622e-36   0.600000   \n",
       "37        48              VraR  1.401629e-24  2.803259e-24   0.450000   \n",
       "38        49              WalR  2.244531e-30  1.571172e-29   0.800000   \n",
       "39        49              SaeR  6.008128e-28  2.102845e-27   0.485714   \n",
       "40        50              NreC  0.000000e+00  0.000000e+00   1.000000   \n",
       "41        50               Rex  5.856495e-25  8.784742e-25   1.000000   \n",
       "42        50              VraR  6.598882e-11  6.598882e-11   0.583333   \n",
       "43        52              ManR  0.000000e+00  0.000000e+00   1.000000   \n",
       "44        54              CodY  4.597413e-15  4.597413e-15   0.736842   \n",
       "45        55               Fur  1.358287e-60  4.074860e-60   0.649123   \n",
       "46        58              LacR  1.326202e-14  1.326202e-14   0.714286   \n",
       "47        59               Rex  1.105581e-17  5.527907e-17   0.900000   \n",
       "48        62        L20_leader  1.440816e-07  5.763265e-07   0.187500   \n",
       "49        63              WalR  2.557458e-17  2.813204e-16   0.385714   \n",
       "50        65              CcpA  4.631188e-17  3.704950e-16   0.714286   \n",
       "51        65              MalR  1.352868e-14  5.411472e-14   0.333333   \n",
       "52        65              TreR  3.421939e-07  9.125170e-07   0.142857   \n",
       "53        66               Fur  2.876360e-15  2.876360e-15   1.000000   \n",
       "54        67              ArgR  9.774854e-09  3.909942e-08   0.666667   \n",
       "55        68              WalR  1.012067e-10  8.096536e-10   0.390244   \n",
       "56        69              CcpA  5.104426e-09  5.104426e-09   1.000000   \n",
       "57        73              CodY  7.090098e-18  1.418020e-17   0.933333   \n",
       "58        73            Lysine  6.104774e-08  6.104774e-08   0.266667   \n",
       "\n",
       "      recall   f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0   1.000000  1.000000   3.0           3.0            3.0     1.0  \n",
       "1   0.076433  0.139535  12.0         157.0           15.0     1.0  \n",
       "2   0.351351  0.495238  26.0          74.0           31.0     1.0  \n",
       "3   0.142857  0.164384   6.0          42.0           31.0     1.0  \n",
       "4   1.000000  0.666667   4.0           4.0            8.0     1.0  \n",
       "5   1.000000  1.000000   7.0           7.0            7.0     1.0  \n",
       "6   0.857143  0.857143  12.0          14.0           14.0     1.0  \n",
       "7   0.133333  0.210526   6.0          45.0           12.0     1.0  \n",
       "8   0.595041  0.623377  72.0         121.0          110.0     1.0  \n",
       "9   1.000000  0.151261   9.0           9.0          110.0     1.0  \n",
       "10  0.833333  0.909091   5.0           6.0            5.0     1.0  \n",
       "11  0.041322  0.078125   5.0         121.0            7.0     1.0  \n",
       "12  0.212121  0.194444   7.0          33.0           39.0     1.0  \n",
       "13  1.000000  0.909091   5.0           5.0            6.0     1.0  \n",
       "14  1.000000  1.000000   5.0           5.0            5.0     1.0  \n",
       "15  0.312500  0.476190   5.0          16.0            5.0     1.0  \n",
       "16  0.151515  0.263158   5.0          33.0            5.0     1.0  \n",
       "17  0.041322  0.079365   5.0         121.0            5.0     1.0  \n",
       "18  1.000000  0.941176   8.0           8.0            9.0     1.0  \n",
       "19  0.038217  0.072289   6.0         157.0            9.0     1.0  \n",
       "20  0.476190  0.540541  10.0          21.0           16.0     1.0  \n",
       "21  0.360544  0.454936  53.0         147.0           86.0     1.0  \n",
       "22  0.777778  0.147368   7.0           9.0           86.0     1.0  \n",
       "23  0.857143  0.129032   6.0           7.0           86.0     1.0  \n",
       "24  0.538462  0.141414   7.0          13.0           86.0     1.0  \n",
       "25  1.000000  0.109890   5.0           5.0           86.0     1.0  \n",
       "26  0.061224  0.115385   9.0         147.0            9.0     1.0  \n",
       "27  0.769231  0.869565  10.0          13.0           10.0     1.0  \n",
       "28  0.833333  0.625000   5.0           6.0           10.0     1.0  \n",
       "29  0.736842  0.717949  14.0          19.0           20.0     1.0  \n",
       "30  0.916667  0.511628  22.0          24.0           62.0     1.0  \n",
       "31  1.000000  0.833333   5.0           5.0            7.0     1.0  \n",
       "32  0.283784  0.432990  21.0          74.0           23.0     1.0  \n",
       "33  1.000000  1.000000   2.0           2.0            2.0     1.0  \n",
       "34  0.148649  0.250000  11.0          74.0           14.0     1.0  \n",
       "35  0.750000  0.750000   3.0           4.0            4.0     1.0  \n",
       "36  0.533333  0.564706  24.0          45.0           40.0     1.0  \n",
       "37  0.428571  0.439024  18.0          42.0           40.0     1.0  \n",
       "38  0.178344  0.291667  28.0         157.0           35.0     1.0  \n",
       "39  0.607143  0.539683  17.0          28.0           35.0     1.0  \n",
       "40  1.000000  1.000000  12.0          12.0           12.0     1.0  \n",
       "41  0.363636  0.533333  12.0          33.0           12.0     1.0  \n",
       "42  0.166667  0.259259   7.0          42.0           12.0     1.0  \n",
       "43  1.000000  1.000000   2.0           2.0            2.0     1.0  \n",
       "44  0.095238  0.168675  14.0         147.0           19.0     1.0  \n",
       "45  0.787234  0.711538  37.0          47.0           57.0     1.0  \n",
       "46  1.000000  0.833333   5.0           5.0            7.0     1.0  \n",
       "47  0.272727  0.418605   9.0          33.0           10.0     1.0  \n",
       "48  1.000000  0.315789   3.0           3.0           16.0     1.0  \n",
       "49  0.171975  0.237885  27.0         157.0           70.0     1.0  \n",
       "50  0.123967  0.211268  15.0         121.0           21.0     1.0  \n",
       "51  0.777778  0.466667   7.0           9.0           21.0     1.0  \n",
       "52  1.000000  0.250000   3.0           3.0           21.0     1.0  \n",
       "53  0.170213  0.290909   8.0          47.0            8.0     1.0  \n",
       "54  0.250000  0.363636   4.0          16.0            6.0     1.0  \n",
       "55  0.101911  0.161616  16.0         157.0           41.0     1.0  \n",
       "56  0.049587  0.094488   6.0         121.0            6.0     1.0  \n",
       "57  0.095238  0.172840  14.0         147.0           15.0     1.0  \n",
       "58  0.444444  0.333333   4.0           9.0           15.0     1.0  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First search for regulator enrichments with 2 regulators\n",
    "ica_data.compute_trn_enrichment(max_regs=2,save=True)\n",
    "\n",
    "# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\n",
    "ica_data.compute_trn_enrichment(max_regs=1,save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of regulatory iModulons are shown below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "76 Total iModulons\n",
      "42 Regulatory iModulons\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FruR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WalR</td>\n",
       "      <td>1.977615e-13</td>\n",
       "      <td>3.955230e-13</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.076433</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>12.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>6.486677e-39</td>\n",
       "      <td>1.946003e-38</td>\n",
       "      <td>0.838710</td>\n",
       "      <td>0.351351</td>\n",
       "      <td>0.495238</td>\n",
       "      <td>26.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Purine</td>\n",
       "      <td>2.523321e-11</td>\n",
       "      <td>7.569963e-11</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TPP</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PurR</td>\n",
       "      <td>1.365061e-29</td>\n",
       "      <td>2.730122e-29</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>VraS2</td>\n",
       "      <td>9.312770e-09</td>\n",
       "      <td>2.793831e-08</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>6.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>1.213330e-81</td>\n",
       "      <td>2.426660e-80</td>\n",
       "      <td>0.654545</td>\n",
       "      <td>0.595041</td>\n",
       "      <td>0.623377</td>\n",
       "      <td>72.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CggR</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>2.460824e-06</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>5.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Rex</td>\n",
       "      <td>1.656345e-07</td>\n",
       "      <td>8.281726e-07</td>\n",
       "      <td>0.179487</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>7.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FMN</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>3.789150e-15</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ArcR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PyrR</td>\n",
       "      <td>8.232349e-23</td>\n",
       "      <td>1.646470e-22</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Genomic Island 3</td>\n",
       "      <td>2.803906e-19</td>\n",
       "      <td>5.607811e-19</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.540541</td>\n",
       "      <td>10.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CodY</td>\n",
       "      <td>3.741945e-50</td>\n",
       "      <td>4.864528e-49</td>\n",
       "      <td>0.616279</td>\n",
       "      <td>0.360544</td>\n",
       "      <td>0.454936</td>\n",
       "      <td>53.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CodY</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.987481e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>9.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CtsR</td>\n",
       "      <td>2.899696e-26</td>\n",
       "      <td>8.699089e-26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SaeR/Genomic Island 8</td>\n",
       "      <td>1.783157e-10</td>\n",
       "      <td>1.429378e-06</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>8.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CymR</td>\n",
       "      <td>1.655744e-29</td>\n",
       "      <td>4.967233e-29</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.717949</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Genomic Island 1</td>\n",
       "      <td>1.040356e-36</td>\n",
       "      <td>3.121069e-36</td>\n",
       "      <td>0.354839</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>22.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>AgrA</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>2.652405e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>5.380175e-33</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>0.283784</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HssR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Genomic Island 9</td>\n",
       "      <td>5.602650e-16</td>\n",
       "      <td>2.241060e-15</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>0.148649</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MntR</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>4.115537e-09</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>VraS2</td>\n",
       "      <td>1.639155e-36</td>\n",
       "      <td>6.556622e-36</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.564706</td>\n",
       "      <td>24.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>WalR</td>\n",
       "      <td>2.244531e-30</td>\n",
       "      <td>1.571172e-29</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.178344</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>28.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NreC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>ManR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CodY</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>4.597413e-15</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.168675</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Fur</td>\n",
       "      <td>1.358287e-60</td>\n",
       "      <td>4.074860e-60</td>\n",
       "      <td>0.649123</td>\n",
       "      <td>0.787234</td>\n",
       "      <td>0.711538</td>\n",
       "      <td>37.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>LacR</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>1.326202e-14</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Rex</td>\n",
       "      <td>1.105581e-17</td>\n",
       "      <td>5.527907e-17</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.418605</td>\n",
       "      <td>9.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>L20_leader</td>\n",
       "      <td>1.440816e-07</td>\n",
       "      <td>5.763265e-07</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>WalR</td>\n",
       "      <td>2.557458e-17</td>\n",
       "      <td>2.813204e-16</td>\n",
       "      <td>0.385714</td>\n",
       "      <td>0.171975</td>\n",
       "      <td>0.237885</td>\n",
       "      <td>27.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>4.631188e-17</td>\n",
       "      <td>3.704950e-16</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.123967</td>\n",
       "      <td>0.211268</td>\n",
       "      <td>15.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Fur</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>2.876360e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>0.290909</td>\n",
       "      <td>8.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>ArgR</td>\n",
       "      <td>9.774854e-09</td>\n",
       "      <td>3.909942e-08</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>WalR</td>\n",
       "      <td>1.012067e-10</td>\n",
       "      <td>8.096536e-10</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0.101911</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>16.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>5.104426e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049587</td>\n",
       "      <td>0.094488</td>\n",
       "      <td>6.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>CodY</td>\n",
       "      <td>7.090098e-18</td>\n",
       "      <td>1.418020e-17</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.172840</td>\n",
       "      <td>14.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                regulator        pvalue        qvalue  precision    recall  \\\n",
       "1                    FruR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "3                    WalR  1.977615e-13  3.955230e-13   0.800000  0.076433   \n",
       "5        Genomic Island 9  6.486677e-39  1.946003e-38   0.838710  0.351351   \n",
       "6                  Purine  2.523321e-11  7.569963e-11   0.500000  1.000000   \n",
       "9                     TPP  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "10                   PurR  1.365061e-29  2.730122e-29   0.857143  0.857143   \n",
       "11                  VraS2  9.312770e-09  2.793831e-08   0.500000  0.133333   \n",
       "13                   CcpA  1.213330e-81  2.426660e-80   0.654545  0.595041   \n",
       "15                   CggR  3.789150e-15  3.789150e-15   1.000000  0.833333   \n",
       "17                   CcpA  2.460824e-06  2.460824e-06   0.714286  0.041322   \n",
       "20                    Rex  1.656345e-07  8.281726e-07   0.179487  0.212121   \n",
       "22                    FMN  3.789150e-15  3.789150e-15   0.833333  1.000000   \n",
       "23                   ArcR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "24                   PyrR  8.232349e-23  1.646470e-22   0.888889  1.000000   \n",
       "28       Genomic Island 3  2.803906e-19  5.607811e-19   0.625000  0.476190   \n",
       "29                   CodY  3.741945e-50  4.864528e-49   0.616279  0.360544   \n",
       "31                   CodY  1.987481e-12  1.987481e-12   1.000000  0.061224   \n",
       "32                   CtsR  2.899696e-26  8.699089e-26   1.000000  0.769231   \n",
       "35  SaeR/Genomic Island 8  1.783157e-10  1.429378e-06   0.400000  0.181818   \n",
       "37                   CymR  1.655744e-29  4.967233e-29   0.700000  0.736842   \n",
       "38       Genomic Island 1  1.040356e-36  3.121069e-36   0.354839  0.916667   \n",
       "39                   AgrA  1.326202e-14  2.652405e-14   0.714286  1.000000   \n",
       "40       Genomic Island 9  5.380175e-33  5.380175e-33   0.913043  0.283784   \n",
       "41                   HssR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "43       Genomic Island 9  5.602650e-16  2.241060e-15   0.785714  0.148649   \n",
       "46                   MntR  4.115537e-09  4.115537e-09   0.750000  0.750000   \n",
       "48                  VraS2  1.639155e-36  6.556622e-36   0.600000  0.533333   \n",
       "49                   WalR  2.244531e-30  1.571172e-29   0.800000  0.178344   \n",
       "50                   NreC  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "52                   ManR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "54                   CodY  4.597413e-15  4.597413e-15   0.736842  0.095238   \n",
       "55                    Fur  1.358287e-60  4.074860e-60   0.649123  0.787234   \n",
       "58                   LacR  1.326202e-14  1.326202e-14   0.714286  1.000000   \n",
       "59                    Rex  1.105581e-17  5.527907e-17   0.900000  0.272727   \n",
       "62             L20_leader  1.440816e-07  5.763265e-07   0.187500  1.000000   \n",
       "63                   WalR  2.557458e-17  2.813204e-16   0.385714  0.171975   \n",
       "65                   CcpA  4.631188e-17  3.704950e-16   0.714286  0.123967   \n",
       "66                    Fur  2.876360e-15  2.876360e-15   1.000000  0.170213   \n",
       "67                   ArgR  9.774854e-09  3.909942e-08   0.666667  0.250000   \n",
       "68                   WalR  1.012067e-10  8.096536e-10   0.390244  0.101911   \n",
       "69                   CcpA  5.104426e-09  5.104426e-09   1.000000  0.049587   \n",
       "73                   CodY  7.090098e-18  1.418020e-17   0.933333  0.095238   \n",
       "\n",
       "     f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "1   1.000000   3.0           3.0            3.0     1.0  \n",
       "3   0.139535  12.0         157.0           15.0     1.0  \n",
       "5   0.495238  26.0          74.0           31.0     1.0  \n",
       "6   0.666667   4.0           4.0            8.0     1.0  \n",
       "9   1.000000   7.0           7.0            7.0     1.0  \n",
       "10  0.857143  12.0          14.0           14.0     1.0  \n",
       "11  0.210526   6.0          45.0           12.0     1.0  \n",
       "13  0.623377  72.0         121.0          110.0     1.0  \n",
       "15  0.909091   5.0           6.0            5.0     1.0  \n",
       "17  0.078125   5.0         121.0            7.0     1.0  \n",
       "20  0.194444   7.0          33.0           39.0     1.0  \n",
       "22  0.909091   5.0           5.0            6.0     1.0  \n",
       "23  1.000000   5.0           5.0            5.0     1.0  \n",
       "24  0.941176   8.0           8.0            9.0     1.0  \n",
       "28  0.540541  10.0          21.0           16.0     1.0  \n",
       "29  0.454936  53.0         147.0           86.0     1.0  \n",
       "31  0.115385   9.0         147.0            9.0     1.0  \n",
       "32  0.869565  10.0          13.0           10.0     1.0  \n",
       "35  0.250000   8.0          44.0           20.0     2.0  \n",
       "37  0.717949  14.0          19.0           20.0     1.0  \n",
       "38  0.511628  22.0          24.0           62.0     1.0  \n",
       "39  0.833333   5.0           5.0            7.0     1.0  \n",
       "40  0.432990  21.0          74.0           23.0     1.0  \n",
       "41  1.000000   2.0           2.0            2.0     1.0  \n",
       "43  0.250000  11.0          74.0           14.0     1.0  \n",
       "46  0.750000   3.0           4.0            4.0     1.0  \n",
       "48  0.564706  24.0          45.0           40.0     1.0  \n",
       "49  0.291667  28.0         157.0           35.0     1.0  \n",
       "50  1.000000  12.0          12.0           12.0     1.0  \n",
       "52  1.000000   2.0           2.0            2.0     1.0  \n",
       "54  0.168675  14.0         147.0           19.0     1.0  \n",
       "55  0.711538  37.0          47.0           57.0     1.0  \n",
       "58  0.833333   5.0           5.0            7.0     1.0  \n",
       "59  0.418605   9.0          33.0           10.0     1.0  \n",
       "62  0.315789   3.0           3.0           16.0     1.0  \n",
       "63  0.237885  27.0         157.0           70.0     1.0  \n",
       "65  0.211268  15.0         121.0           21.0     1.0  \n",
       "66  0.290909   8.0          47.0            8.0     1.0  \n",
       "67  0.363636   4.0          16.0            6.0     1.0  \n",
       "68  0.161616  16.0         157.0           41.0     1.0  \n",
       "69  0.094488   6.0         121.0            6.0     1.0  \n",
       "73  0.172840  14.0         147.0           15.0     1.0  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]\n",
    "print(len(ica_data.imodulon_table),'Total iModulons')\n",
    "print(len(regulatory_imodulons),'Regulatory iModulons')\n",
    "regulatory_imodulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can rename iModulons in this jupyter notebook, or you can save the iModulon table as a CSV and edit it in Excel.\n",
    "\n",
    "If two iModulons have the same regulator (e.g. 'Reg'), they will be named 'Reg-1' and 'Reg-2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Duplicate iModulon names detected. iModulon 3 will be renamed to WalR-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 5 will be renamed to Genomic Island 9-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 11 will be renamed to VraS2-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 13 will be renamed to CcpA-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 17 will be renamed to CcpA-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 20 will be renamed to Rex-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 29 will be renamed to CodY-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 31 will be renamed to CodY-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 40 will be renamed to Genomic Island 9-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 43 will be renamed to Genomic Island 9-3\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 48 will be renamed to VraS2-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 49 will be renamed to WalR-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 54 will be renamed to CodY-3\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 55 will be renamed to Fur-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 59 will be renamed to Rex-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 63 will be renamed to WalR-3\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 65 will be renamed to CcpA-3\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 66 will be renamed to Fur-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 68 will be renamed to WalR-4\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 69 will be renamed to CcpA-4\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 73 will be renamed to CodY-4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FruR</th>\n",
       "      <td>FruR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WalR-1</th>\n",
       "      <td>WalR</td>\n",
       "      <td>1.977615e-13</td>\n",
       "      <td>3.955230e-13</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.076433</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>12.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       regulator        pvalue        qvalue  precision    recall   f1score  \\\n",
       "0            NaN           NaN           NaN        NaN       NaN       NaN   \n",
       "FruR        FruR  0.000000e+00  0.000000e+00        1.0  1.000000  1.000000   \n",
       "2            NaN           NaN           NaN        NaN       NaN       NaN   \n",
       "WalR-1      WalR  1.977615e-13  3.955230e-13        0.8  0.076433  0.139535   \n",
       "4            NaN           NaN           NaN        NaN       NaN       NaN   \n",
       "\n",
       "          TP  regulon_size  imodulon_size  n_regs  \n",
       "0        NaN           NaN            NaN     NaN  \n",
       "FruR     3.0           3.0            3.0     1.0  \n",
       "2        NaN           NaN            NaN     NaN  \n",
       "WalR-1  12.0         157.0           15.0     1.0  \n",
       "4        NaN           NaN            NaN     NaN  "
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.rename_imodulons(regulatory_imodulons.regulator.to_dict())\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO = pd.read_csv(path.join(work_dir, 'external', 'eggNOG_annotations.txt'), sep='\\t', skiprows=4)\n",
    "DF_GO = DF_GO.rename(columns={'#query': 'gene_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'COG_category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>E</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.051398</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.032967</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>6.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WalR-1</td>\n",
       "      <td>M</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.002784</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.042553</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>6.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genomic Island 9-1</td>\n",
       "      <td>S</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.013219</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>0.025890</td>\n",
       "      <td>0.049307</td>\n",
       "      <td>16.0</td>\n",
       "      <td>618.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Purine</td>\n",
       "      <td>F</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>5.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TPP</td>\n",
       "      <td>H</td>\n",
       "      <td>0.000760</td>\n",
       "      <td>0.047127</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.036145</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>3.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             imodulon COG_category    pvalue    qvalue  precision    recall  \\\n",
       "0                   0            E  0.000829  0.051398   0.315789  0.032967   \n",
       "1              WalR-1            M  0.000045  0.002784   0.400000  0.042553   \n",
       "2  Genomic Island 9-1            S  0.000213  0.013219   0.516129  0.025890   \n",
       "3              Purine            F  0.000002  0.000117   0.625000  0.052632   \n",
       "4                 TPP            H  0.000760  0.047127   0.428571  0.036145   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.059701   6.0            182.0           19.0  \n",
       "1  0.076923   6.0            141.0           15.0  \n",
       "2  0.049307  16.0            618.0           31.0  \n",
       "3  0.097087   5.0             95.0            8.0  \n",
       "4  0.066667   3.0             83.0            7.0  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_ko' 'KEGG_Pathway' 'KEGG_Module' 'KEGG_Reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>USA300HOU_RS00050</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>USA300HOU_RS00055</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>USA300HOU_RS00065</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>USA300HOU_RS00070</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>USA300HOU_RS00110</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id database kegg_id\n",
       "10  USA300HOU_RS00050  KEGG_ko       -\n",
       "11  USA300HOU_RS00055  KEGG_ko       -\n",
       "13  USA300HOU_RS00065  KEGG_ko       -\n",
       "14  USA300HOU_RS00070  KEGG_ko       -\n",
       "20  USA300HOU_RS00110  KEGG_ko       -"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_Pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_Module']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map00791</td>\n",
       "      <td>2.493127e-07</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>map00220</td>\n",
       "      <td>3.073765e-04</td>\n",
       "      <td>0.030584</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FruR</td>\n",
       "      <td>map00051</td>\n",
       "      <td>9.958511e-05</td>\n",
       "      <td>0.019817</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>map00260</td>\n",
       "      <td>3.619166e-04</td>\n",
       "      <td>0.072021</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>2.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Purine</td>\n",
       "      <td>map00230</td>\n",
       "      <td>1.695523e-05</td>\n",
       "      <td>0.003374</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>4.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon   kegg_id        pvalue    qvalue  precision    recall   f1score  \\\n",
       "0        0  map00791  2.493127e-07  0.000050   0.157895  1.000000  0.272727   \n",
       "1        0  map00220  3.073765e-04  0.030584   0.157895  0.142857  0.150000   \n",
       "2     FruR  map00051  9.958511e-05  0.019817   0.666667  0.117647  0.200000   \n",
       "3        4  map00260  3.619166e-04  0.072021   0.666667  0.062500  0.114286   \n",
       "4   Purine  map00230  1.695523e-05  0.003374   0.500000  0.060606  0.108108   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \n",
       "0  3.0              3.0           19.0  \n",
       "1  3.0             21.0           19.0  \n",
       "2  2.0             17.0            3.0  \n",
       "3  2.0             32.0            3.0  \n",
       "4  4.0             66.0            8.0  "
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d7c3da2bddbc434fb097892481f00919",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/150 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "73cf991ea2004943a1d00bf3fc137b18",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/114 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    text = REST.kegg_find('pathway',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    text = REST.kegg_find('module',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map00791</td>\n",
       "      <td>2.493127e-07</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Atrazine degradation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>map00220</td>\n",
       "      <td>3.073765e-04</td>\n",
       "      <td>0.030584</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Arginine biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FruR</td>\n",
       "      <td>map00051</td>\n",
       "      <td>9.958511e-05</td>\n",
       "      <td>0.019817</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Fructose and mannose metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>map00260</td>\n",
       "      <td>3.619166e-04</td>\n",
       "      <td>0.072021</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>2.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Purine</td>\n",
       "      <td>map00230</td>\n",
       "      <td>1.695523e-05</td>\n",
       "      <td>0.003374</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>4.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>Purine metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon   kegg_id        pvalue    qvalue  precision    recall   f1score  \\\n",
       "0        0  map00791  2.493127e-07  0.000050   0.157895  1.000000  0.272727   \n",
       "1        0  map00220  3.073765e-04  0.030584   0.157895  0.142857  0.150000   \n",
       "2     FruR  map00051  9.958511e-05  0.019817   0.666667  0.117647  0.200000   \n",
       "3        4  map00260  3.619166e-04  0.072021   0.666667  0.062500  0.114286   \n",
       "4   Purine  map00230  1.695523e-05  0.003374   0.500000  0.060606  0.108108   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \\\n",
       "0  3.0              3.0           19.0   \n",
       "1  3.0             21.0           19.0   \n",
       "2  2.0             17.0            3.0   \n",
       "3  2.0             32.0            3.0   \n",
       "4  4.0             66.0            8.0   \n",
       "\n",
       "                               pathway_name  \n",
       "0                      Atrazine degradation  \n",
       "1                     Arginine biosynthesis  \n",
       "2           Fructose and mannose metabolism  \n",
       "3  Glycine, serine and threonine metabolism  \n",
       "4                         Purine metabolism  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>M00555</td>\n",
       "      <td>7.348163e-07</td>\n",
       "      <td>1.873782e-04</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Betaine biosynthesis, choline =&gt; betaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Purine</td>\n",
       "      <td>M00050</td>\n",
       "      <td>6.829507e-05</td>\n",
       "      <td>1.741524e-02</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>Guanine ribonucleotide biosynthesis, IMP =&gt; GD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>M00725</td>\n",
       "      <td>8.801943e-10</td>\n",
       "      <td>2.244495e-07</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Cationic antimicrobial peptide (CAMP) resistan...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TPP</td>\n",
       "      <td>M00127</td>\n",
       "      <td>9.005102e-09</td>\n",
       "      <td>2.296301e-06</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Thiamine biosynthesis, prokaryotes, AIR (+ DXP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TPP</td>\n",
       "      <td>M00582</td>\n",
       "      <td>2.548499e-06</td>\n",
       "      <td>3.249336e-04</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon kegg_id        pvalue        qvalue  precision    recall   f1score  \\\n",
       "0        4  M00555  7.348163e-07  1.873782e-04   0.666667  1.000000  0.800000   \n",
       "1   Purine  M00050  6.829507e-05  1.741524e-02   0.250000  0.400000  0.307692   \n",
       "2        7  M00725  8.801943e-10  2.244495e-07   0.571429  0.500000  0.533333   \n",
       "3      TPP  M00127  9.005102e-09  2.296301e-06   0.428571  1.000000  0.600000   \n",
       "4      TPP  M00582  2.548499e-06  3.249336e-04   0.428571  0.230769  0.300000   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \\\n",
       "0  2.0              2.0            3.0   \n",
       "1  2.0              5.0            8.0   \n",
       "2  4.0              8.0            7.0   \n",
       "3  3.0              3.0            7.0   \n",
       "4  3.0             13.0            7.0   \n",
       "\n",
       "                                         module_name  \n",
       "0           Betaine biosynthesis, choline => betaine  \n",
       "1  Guanine ribonucleotide biosynthesis, IMP => GD...  \n",
       "2  Cationic antimicrobial peptide (CAMP) resistan...  \n",
       "3  Thiamine biosynthesis, prokaryotes, AIR (+ DXP...  \n",
       "4                                               None  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'EggNOG'\n",
    "DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "DF_module_enrich['source'] = 'KEGG modules'\n",
    "\n",
    "DF_GO_enrich.rename({'COG_category':'annotation'},axis=1, inplace=True)\n",
    "DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_pathway_enrich, DF_module_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SG_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator  pvalue  qvalue  precision  recall  f1score  TP  regulon_size  \\\n",
       "SG_1       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_2       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_3       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_4       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_5       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_6       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_7       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_8       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "\n",
       "      imodulon_size  n_regs single_gene  \n",
       "SG_1            NaN     NaN        True  \n",
       "SG_2            NaN     NaN        True  \n",
       "SG_3            NaN     NaN        True  \n",
       "SG_4            NaN     NaN        True  \n",
       "SG_5            NaN     NaN        True  \n",
       "SG_6            NaN     NaN        True  \n",
       "SG_7            NaN     NaN        True  \n",
       "SG_8            NaN     NaN        True  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join(work_dir,'interim','saureus_raw.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join(data_dir,'imodulon_table_raw.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
