{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.io import *\n",
    "from pymodulon.motif import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_json_model(os.path.join('/home/jovyan/work/','processed_data','saureus.json.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biocyc_annotations.txt\tgenome.gff3\t\t  plasmid_pUSA300HOUMR.fasta\r\n",
      "CDS.fna\t\t\tkegg_mapping.csv\t  plasmid_pUSA300HOUMR.gff3\r\n",
      "eggNOG_annotations.txt\tplasmid_pUSA01-HOU.fasta  TRN.csv\r\n",
      "genome.fasta\t\tplasmid_pUSA01-HOU.gff3\r\n"
     ]
    }
   ],
   "source": [
    "!ls /home/jovyan/work/external/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = [os.path.join('/home/jovyan/work/','external','genome.fasta'),\n",
    "             os.path.join('/home/jovyan/work/','external','plasmid_pUSA01-HOU.fasta'),\n",
    "             os.path.join('/home/jovyan/work/','external','plasmid_pUSA300HOUMR.fasta')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search for motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/LacR.fasta -oc motifs/LacR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 1 motif across 3 sites\n",
      "Finding motifs for 28 sequences\n",
      "Running command: meme motifs/Rex-1.fasta -oc motifs/Rex-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 9\n",
      "Found 1 motif across 11 sites\n",
      "Finding motifs for 21 sequences\n",
      "Running command: meme motifs/SarA.fasta -oc motifs/SarA -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 7\n",
      "Found 5 motifs across 37 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/Purine.fasta -oc motifs/Purine -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/SG_1.fasta -oc motifs/SG_1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 17 sequences\n",
      "Running command: meme motifs/Genomic_Island_9-1.fasta -oc motifs/Genomic_Island_9-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 5\n",
      "Found 3 motifs across 23 sites\n",
      "Finding motifs for 9 sequences\n",
      "Running command: meme motifs/SG_2.fasta -oc motifs/SG_2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/ArgR.fasta -oc motifs/ArgR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 7 sequences\n",
      "Running command: meme motifs/Phenol_Soluble_Modulin.fasta -oc motifs/Phenol_Soluble_Modulin -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 5 motifs across 29 sites\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/NreC.fasta -oc motifs/NreC -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 6 sequences\n",
      "Running command: meme motifs/CodY-1.fasta -oc motifs/CodY-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 2 motifs across 6 sites\n",
      "Finding motifs for 16 sequences\n",
      "Running command: meme motifs/scfm2-artefact.fasta -oc motifs/scfm2-artefact -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 5\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 31 sequences\n",
      "Running command: meme motifs/SaeR-1.fasta -oc motifs/SaeR-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 10\n",
      "Found 4 motifs across 82 sites\n",
      "Finding motifs for 46 sequences\n",
      "Running command: meme motifs/SigS-1.fasta -oc motifs/SigS-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 15\n",
      "Found 1 motif across 46 sites\n",
      "Less than two sequences found for iModulon: HssR\n",
      "Finding motifs for 8 sequences\n",
      "Running command: meme motifs/SG_3.fasta -oc motifs/SG_3 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 22 sequences\n",
      "Running command: meme motifs/Uncharacterized-1.fasta -oc motifs/Uncharacterized-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 7\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/TPP.fasta -oc motifs/TPP -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/Beta-Lactam_Resistance.fasta -oc motifs/Beta-Lactam_Resistance -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 5 motifs across 22 sites\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/SG_4.fasta -oc motifs/SG_4 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "No motif found with E-value < 1.0e-03\n",
      "Less than two sequences found for iModulon: FruR\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/FMN.fasta -oc motifs/FMN -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 1 motif across 3 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/CggR.fasta -oc motifs/CggR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 9 sequences\n",
      "Running command: meme motifs/Genomic_Island_9-2.fasta -oc motifs/Genomic_Island_9-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "Found 3 motifs across 14 sites\n",
      "Finding motifs for 7 sequences\n",
      "Running command: meme motifs/SG_5.fasta -oc motifs/SG_5 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 45 sequences\n",
      "Running command: meme motifs/CodY-2.fasta -oc motifs/CodY-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 15\n",
      "Found 3 motifs across 63 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/Fur-1.fasta -oc motifs/Fur-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 12 sequences\n",
      "Running command: meme motifs/PerR.fasta -oc motifs/PerR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "Found 2 motifs across 10 sites\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/SG_6.fasta -oc motifs/SG_6 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 18 sequences\n",
      "Running command: meme motifs/Translation.fasta -oc motifs/Translation -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 6\n",
      "Found 3 motifs across 36 sites\n",
      "Finding motifs for 17 sequences\n",
      "Running command: meme motifs/SaeR-2.fasta -oc motifs/SaeR-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 5\n",
      "Found 5 motifs across 37 sites\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/SG_7.fasta -oc motifs/SG_7 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/CAMP_Resistance.fasta -oc motifs/CAMP_Resistance -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 2 motifs across 5 sites\n",
      "Finding motifs for 14 sequences\n",
      "Running command: meme motifs/Vir-1.fasta -oc motifs/Vir-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "Found 1 motif across 7 sites\n",
      "Finding motifs for 40 sequences\n",
      "Running command: meme motifs/SigS-2.fasta -oc motifs/SigS-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 13\n",
      "Found 1 motif across 13 sites\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/Genomic_Island_9-3.fasta -oc motifs/Genomic_Island_9-3 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "Found 5 motifs across 33 sites\n",
      "Finding motifs for 6 sequences\n",
      "Running command: meme motifs/SG_9.fasta -oc motifs/SG_9 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 8 sequences\n",
      "Running command: meme motifs/Genomic_Island_3.fasta -oc motifs/Genomic_Island_3 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 1 motif across 7 sites\n",
      "Finding motifs for 27 sequences\n",
      "Running command: meme motifs/VraS2-1.fasta -oc motifs/VraS2-1 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 9\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 11 sequences\n",
      "Running command: meme motifs/CymR.fasta -oc motifs/CymR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 38 sequences\n",
      "Running command: meme motifs/SigB.fasta -oc motifs/SigB -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 12\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/PurR.fasta -oc motifs/PurR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No motif found with E-value < 1.0e-03\n",
      "Less than two sequences found for iModulon: TA system\n",
      "Finding motifs for 38 sequences\n",
      "Running command: meme motifs/ArlRS_MgrA.fasta -oc motifs/ArlRS_MgrA -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 12\n",
      "Found 2 motifs across 29 sites\n",
      "Less than two sequences found for iModulon: CodY/ILVopr\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/Glycerol_Metabolism.fasta -oc motifs/Glycerol_Metabolism -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 7 sequences\n",
      "Running command: meme motifs/CtsR.fasta -oc motifs/CtsR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 2 motifs across 9 sites\n",
      "Finding motifs for 13 sequences\n",
      "Running command: meme motifs/Autolysin.fasta -oc motifs/Autolysin -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "Found 1 motif across 13 sites\n",
      "Less than two sequences found for iModulon: lrgAB\n",
      "Finding motifs for 28 sequences\n",
      "Running command: meme motifs/pUSA300HOUMR.fasta -oc motifs/pUSA300HOUMR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 9\n",
      "Found 5 motifs across 55 sites\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/Vir-2.fasta -oc motifs/Vir-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 28 sequences\n",
      "Running command: meme motifs/Fur-2.fasta -oc motifs/Fur-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 9\n",
      "Found 3 motifs across 48 sites\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/AgrA.fasta -oc motifs/AgrA -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 5 motifs across 22 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/ArlRS_KO.fasta -oc motifs/ArlRS_KO -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "Found 3 motifs across 7 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/ArcR.fasta -oc motifs/ArcR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/PurR_Accessory.fasta -oc motifs/PurR_Accessory -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/SG_10.fasta -oc motifs/SG_10 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Less than two sequences found for iModulon: ManR\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/FadR.fasta -oc motifs/FadR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/Phosphotransferase_system.fasta -oc motifs/Phosphotransferase_system -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/Rex-2.fasta -oc motifs/Rex-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "Found 1 motif across 10 sites\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/Betaine_Biosynth..fasta -oc motifs/Betaine_Biosynth. -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/SG_11.fasta -oc motifs/SG_11 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/MntR.fasta -oc motifs/MntR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/Accessory_sec.fasta -oc motifs/Accessory_sec -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 12 sequences\n",
      "Running command: meme motifs/VraS2-2.fasta -oc motifs/VraS2-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "Found 4 motifs across 18 sites\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/SG_12.fasta -oc motifs/SG_12 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 24 sequences\n",
      "Running command: meme motifs/Uncharacterized-2.fasta -oc motifs/Uncharacterized-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 8\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/PyrR.fasta -oc motifs/PyrR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 78 sequences\n",
      "Running command: meme motifs/Uncharacterized-3.fasta -oc motifs/Uncharacterized-3 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 26\n",
      "Found 1 motif across 36 sites\n",
      "Finding motifs for 6 sequences\n",
      "Running command: meme motifs/Phosphate_acquisition.fasta -oc motifs/Phosphate_acquisition -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 12 sequences\n",
      "Running command: meme motifs/CcpA-2.fasta -oc motifs/CcpA-2 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 7 sequences\n",
      "Running command: meme motifs/SG_13.fasta -oc motifs/SG_13 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 20 sequences\n",
      "Running command: meme motifs/Itaconate_response.fasta -oc motifs/Itaconate_response -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 6\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 3 sequences\n",
      "Running command: meme motifs/SG_14.fasta -oc motifs/SG_14 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 69 sequences\n",
      "Running command: meme motifs/CcpA-3.fasta -oc motifs/CcpA-3 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 23\n",
      "Found 2 motifs across 114 sites\n",
      "Finding motifs for 23 sequences\n",
      "Running command: meme motifs/Uncharacterized-4.fasta -oc motifs/Uncharacterized-4 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 7\n",
      "Found 1 motif across 12 sites\n",
      "Finding motifs for 4 sequences\n",
      "Running command: meme motifs/SG_15.fasta -oc motifs/SG_15 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 5 sequences\n",
      "Running command: meme motifs/SG_16.fasta -oc motifs/SG_16 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 8 sequences\n",
      "Running command: meme motifs/SG_17.fasta -oc motifs/SG_17 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 6 sequences\n",
      "Running command: meme motifs/GntR.fasta -oc motifs/GntR -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 2 sequences\n",
      "Running command: meme motifs/SG_18.fasta -oc motifs/SG_18 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 2\n",
      "No motif found with E-value < 1.0e-03\n",
      "Less than two sequences found for iModulon: Uncharacterized-5\n",
      "Finding motifs for 14 sequences\n",
      "Running command: meme motifs/SG_19.fasta -oc motifs/SG_19 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 4\n",
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/SG_20.fasta -oc motifs/SG_20 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No motif found with E-value < 1.0e-03\n",
      "Finding motifs for 10 sequences\n",
      "Running command: meme motifs/Uncharacterized-6.fasta -oc motifs/Uncharacterized-6 -dna -mod zoops -p 8 -nmotifs 5 -evt 0.001 -minw 6 -maxw 40 -allw -minsites 3\n",
      "No motif found with E-value < 1.0e-03\n"
     ]
    }
   ],
   "source": [
    "for k in ica_data.imodulon_names:\n",
    "    find_motifs(ica_data, k, fasta_file, force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'LacR': <MotifInfo with 1 motif across 3 sites>,\n",
       " 'Rex-1': <MotifInfo with 1 motif across 11 sites>,\n",
       " 'SarA': <MotifInfo with 5 motifs across 37 sites>,\n",
       " 'Genomic Island 9-1': <MotifInfo with 3 motifs across 23 sites>,\n",
       " 'Phenol Soluble Modulin': <MotifInfo with 5 motifs across 29 sites>,\n",
       " 'CodY-1': <MotifInfo with 2 motifs across 6 sites>,\n",
       " 'SaeR-1': <MotifInfo with 4 motifs across 82 sites>,\n",
       " 'SigS-1': <MotifInfo with 1 motif across 46 sites>,\n",
       " 'Beta-Lactam Resistance': <MotifInfo with 5 motifs across 22 sites>,\n",
       " 'FMN': <MotifInfo with 1 motif across 3 sites>,\n",
       " 'Genomic Island 9-2': <MotifInfo with 3 motifs across 14 sites>,\n",
       " 'CodY-2': <MotifInfo with 3 motifs across 63 sites>,\n",
       " 'PerR': <MotifInfo with 2 motifs across 10 sites>,\n",
       " 'Translation': <MotifInfo with 3 motifs across 36 sites>,\n",
       " 'SaeR-2': <MotifInfo with 5 motifs across 37 sites>,\n",
       " 'CAMP Resistance': <MotifInfo with 2 motifs across 5 sites>,\n",
       " 'Vir-1': <MotifInfo with 1 motif across 7 sites>,\n",
       " 'SigS-2': <MotifInfo with 1 motif across 13 sites>,\n",
       " 'Genomic Island 9-3': <MotifInfo with 5 motifs across 33 sites>,\n",
       " 'Genomic Island 3': <MotifInfo with 1 motif across 7 sites>,\n",
       " 'ArlRS/MgrA': <MotifInfo with 2 motifs across 29 sites>,\n",
       " 'CtsR': <MotifInfo with 2 motifs across 9 sites>,\n",
       " 'Autolysin': <MotifInfo with 1 motif across 13 sites>,\n",
       " 'pUSA300HOUMR': <MotifInfo with 5 motifs across 55 sites>,\n",
       " 'Fur-2': <MotifInfo with 3 motifs across 48 sites>,\n",
       " 'AgrA': <MotifInfo with 5 motifs across 22 sites>,\n",
       " 'ArlRS KO': <MotifInfo with 3 motifs across 7 sites>,\n",
       " 'Rex-2': <MotifInfo with 1 motif across 10 sites>,\n",
       " 'VraS2-2': <MotifInfo with 4 motifs across 18 sites>,\n",
       " 'Uncharacterized-3': <MotifInfo with 1 motif across 36 sites>,\n",
       " 'CcpA-3': <MotifInfo with 2 motifs across 114 sites>,\n",
       " 'Uncharacterized-4': <MotifInfo with 1 motif across 12 sites>}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.motif_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare motifs to known databases using TOMTOM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: tomtom -oc motifs/LacR/tomtom -thresh 0.001 -incomplete-scores -png motifs/LacR/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Rex-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/Rex-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/SarA/tomtom -thresh 0.001 -incomplete-scores -png motifs/SarA/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Genomic_Island_9-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/Genomic_Island_9-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Phenol_Soluble_Modulin/tomtom -thresh 0.001 -incomplete-scores -png motifs/Phenol_Soluble_Modulin/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/CodY-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/CodY-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/SaeR-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/SaeR-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/SigS-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/SigS-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Beta-Lactam_Resistance/tomtom -thresh 0.001 -incomplete-scores -png motifs/Beta-Lactam_Resistance/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/FMN/tomtom -thresh 0.001 -incomplete-scores -png motifs/FMN/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Genomic_Island_9-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/Genomic_Island_9-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/CodY-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/CodY-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/PerR/tomtom -thresh 0.001 -incomplete-scores -png motifs/PerR/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Translation/tomtom -thresh 0.001 -incomplete-scores -png motifs/Translation/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/SaeR-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/SaeR-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/CAMP_Resistance/tomtom -thresh 0.001 -incomplete-scores -png motifs/CAMP_Resistance/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Vir-1/tomtom -thresh 0.001 -incomplete-scores -png motifs/Vir-1/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: tomtom -oc motifs/SigS-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/SigS-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Genomic_Island_9-3/tomtom -thresh 0.001 -incomplete-scores -png motifs/Genomic_Island_9-3/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Genomic_Island_3/tomtom -thresh 0.001 -incomplete-scores -png motifs/Genomic_Island_3/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/ArlRS_MgrA/tomtom -thresh 0.001 -incomplete-scores -png motifs/ArlRS_MgrA/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/CtsR/tomtom -thresh 0.001 -incomplete-scores -png motifs/CtsR/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Autolysin/tomtom -thresh 0.001 -incomplete-scores -png motifs/Autolysin/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/pUSA300HOUMR/tomtom -thresh 0.001 -incomplete-scores -png motifs/pUSA300HOUMR/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Fur-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/Fur-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/AgrA/tomtom -thresh 0.001 -incomplete-scores -png motifs/AgrA/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/ArlRS_KO/tomtom -thresh 0.001 -incomplete-scores -png motifs/ArlRS_KO/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Rex-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/Rex-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/VraS2-2/tomtom -thresh 0.001 -incomplete-scores -png motifs/VraS2-2/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Uncharacterized-3/tomtom -thresh 0.001 -incomplete-scores -png motifs/Uncharacterized-3/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/CcpA-3/tomtom -thresh 0.001 -incomplete-scores -png motifs/CcpA-3/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n",
      "Running command: tomtom -oc motifs/Uncharacterized-4/tomtom -thresh 0.001 -incomplete-scores -png motifs/Uncharacterized-4/meme.txt /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/collectf.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/dpinteract.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/prodoric.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/regtransbase.meme /opt/conda/lib/python3.9/site-packages/pymodulon/data/meme/SwissRegulon_e_coli.meme\n"
     ]
    }
   ],
   "source": [
    "for info in ica_data.motif_info.values():\n",
    "    compare_motifs(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LacR\n",
      "SaeR-1\n",
      "SigS-1\n",
      "CodY-2\n",
      "SigS-2\n",
      "Genomic Island 3\n",
      "ArlRS/MgrA\n",
      "Fur-2\n",
      "Rex-2\n",
      "Uncharacterized-3\n",
      "CcpA-3\n"
     ]
    }
   ],
   "source": [
    "for k in ica_data.motif_info.keys():\n",
    "    if not ica_data.motif_info[k].matches.empty:\n",
    "        print(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SigB-1</th>\n",
       "      <td>SigB</td>\n",
       "      <td>4.367180e-66</td>\n",
       "      <td>6.987488e-65</td>\n",
       "      <td>0.962264</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.372263</td>\n",
       "      <td>51.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CodY</th>\n",
       "      <td>CodY</td>\n",
       "      <td>2.915811e-73</td>\n",
       "      <td>6.997947e-72</td>\n",
       "      <td>0.821918</td>\n",
       "      <td>0.320856</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>60.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SigW</th>\n",
       "      <td>SigW</td>\n",
       "      <td>8.102748e-84</td>\n",
       "      <td>7.292473e-83</td>\n",
       "      <td>0.872727</td>\n",
       "      <td>0.607595</td>\n",
       "      <td>0.716418</td>\n",
       "      <td>48.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CcpA-1</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>7.893839e-56</td>\n",
       "      <td>2.683905e-54</td>\n",
       "      <td>0.719512</td>\n",
       "      <td>0.225191</td>\n",
       "      <td>0.343023</td>\n",
       "      <td>59.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fur</th>\n",
       "      <td>Fur</td>\n",
       "      <td>1.213043e-74</td>\n",
       "      <td>1.334347e-73</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.622951</td>\n",
       "      <td>0.752475</td>\n",
       "      <td>38.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       regulator        pvalue        qvalue  precision    recall   f1score  \\\n",
       "SigB-1      SigB  4.367180e-66  6.987488e-65   0.962264  0.230769  0.372263   \n",
       "CodY        CodY  2.915811e-73  6.997947e-72   0.821918  0.320856  0.461538   \n",
       "SigW        SigW  8.102748e-84  7.292473e-83   0.872727  0.607595  0.716418   \n",
       "CcpA-1      CcpA  7.893839e-56  2.683905e-54   0.719512  0.225191  0.343023   \n",
       "Fur          Fur  1.213043e-74  1.334347e-73   0.950000  0.622951  0.752475   \n",
       "\n",
       "          TP  regulon_size  imodulon_size  n_regs  single_gene  \n",
       "SigB-1  51.0         221.0           53.0     1.0          NaN  \n",
       "CodY    60.0         187.0           73.0     1.0          NaN  \n",
       "SigW    48.0          79.0           55.0     1.0          NaN  \n",
       "CcpA-1  59.0         262.0           82.0     1.0          NaN  \n",
       "Fur     38.0          61.0           40.0     1.0          NaN  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.loc[['SigB-1','CodY','SigW','CcpA-1','Fur']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>target</th>\n",
       "      <th>target_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Evalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>overlap</th>\n",
       "      <th>optimal_offset</th>\n",
       "      <th>orientation</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>OxyR_P.aeruginosa</td>\n",
       "      <td>EXPREG_00001560</td>\n",
       "      <td>2.119520e-07</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>35</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>collectf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>PhoP</td>\n",
       "      <td>MX000057</td>\n",
       "      <td>5.107030e-07</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>31</td>\n",
       "      <td>-4</td>\n",
       "      <td>-</td>\n",
       "      <td>prodoric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>IHF_P.putida</td>\n",
       "      <td>EXPREG_000006f0</td>\n",
       "      <td>1.672050e-06</td>\n",
       "      <td>0.000988</td>\n",
       "      <td>0.000356</td>\n",
       "      <td>26</td>\n",
       "      <td>-10</td>\n",
       "      <td>+</td>\n",
       "      <td>collectf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>ihf</td>\n",
       "      <td>ihf</td>\n",
       "      <td>2.922770e-06</td>\n",
       "      <td>0.001727</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>34</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>AmrZ_P.aeruginosa</td>\n",
       "      <td>EXPREG_000004d0</td>\n",
       "      <td>6.525520e-06</td>\n",
       "      <td>0.003857</td>\n",
       "      <td>0.000992</td>\n",
       "      <td>24</td>\n",
       "      <td>-9</td>\n",
       "      <td>+</td>\n",
       "      <td>collectf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    motif             target        target_id        pvalue    Evalue  \\\n",
       "0  MEME-1  OxyR_P.aeruginosa  EXPREG_00001560  2.119520e-07  0.000125   \n",
       "1  MEME-1               PhoP         MX000057  5.107030e-07  0.000302   \n",
       "2  MEME-1       IHF_P.putida  EXPREG_000006f0  1.672050e-06  0.000988   \n",
       "3  MEME-1                ihf              ihf  2.922770e-06  0.001727   \n",
       "4  MEME-1  AmrZ_P.aeruginosa  EXPREG_000004d0  6.525520e-06  0.003857   \n",
       "\n",
       "     qvalue  overlap  optimal_offset orientation    database  \n",
       "0  0.000149       35               5           +    collectf  \n",
       "1  0.000149       31              -4           -    prodoric  \n",
       "2  0.000356       26             -10           +    collectf  \n",
       "3  0.000519       34              14           -  dpinteract  \n",
       "4  0.000992       24              -9           +    collectf  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.motif_info['Uncharacterized-3'].matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>target</th>\n",
       "      <th>target_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Evalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>overlap</th>\n",
       "      <th>optimal_offset</th>\n",
       "      <th>orientation</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>ihf</td>\n",
       "      <td>ihf</td>\n",
       "      <td>5.133910e-09</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>+</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>OxyR_P.aeruginosa</td>\n",
       "      <td>EXPREG_00001560</td>\n",
       "      <td>1.568880e-07</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>35</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>collectf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>rpoD17</td>\n",
       "      <td>rpoD17</td>\n",
       "      <td>4.909700e-07</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>25</td>\n",
       "      <td>-12</td>\n",
       "      <td>-</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>RpoN</td>\n",
       "      <td>MX000100</td>\n",
       "      <td>1.036980e-06</td>\n",
       "      <td>0.000613</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>24</td>\n",
       "      <td>-13</td>\n",
       "      <td>+</td>\n",
       "      <td>prodoric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>PhoP</td>\n",
       "      <td>MX000057</td>\n",
       "      <td>1.156850e-06</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>31</td>\n",
       "      <td>-3</td>\n",
       "      <td>+</td>\n",
       "      <td>prodoric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>OxyR</td>\n",
       "      <td>MX000154</td>\n",
       "      <td>1.986300e-06</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>prodoric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>rpoD16</td>\n",
       "      <td>rpoD16</td>\n",
       "      <td>4.630600e-06</td>\n",
       "      <td>0.002737</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>19</td>\n",
       "      <td>-18</td>\n",
       "      <td>-</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>rpoS17</td>\n",
       "      <td>rpoS17</td>\n",
       "      <td>5.842440e-06</td>\n",
       "      <td>0.003453</td>\n",
       "      <td>0.000569</td>\n",
       "      <td>29</td>\n",
       "      <td>-6</td>\n",
       "      <td>+</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>SigE</td>\n",
       "      <td>MX000068</td>\n",
       "      <td>7.194000e-06</td>\n",
       "      <td>0.004252</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>26</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>prodoric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MEME-1</td>\n",
       "      <td>rpoD19</td>\n",
       "      <td>rpoD19</td>\n",
       "      <td>1.156250e-05</td>\n",
       "      <td>0.006833</td>\n",
       "      <td>0.000885</td>\n",
       "      <td>22</td>\n",
       "      <td>-15</td>\n",
       "      <td>-</td>\n",
       "      <td>dpinteract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MEME-2</td>\n",
       "      <td>ArcA_28-0</td>\n",
       "      <td>ArcA_28-0</td>\n",
       "      <td>6.043430e-08</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>SwissRegulon_e_coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     motif             target        target_id        pvalue    Evalue  \\\n",
       "0   MEME-1                ihf              ihf  5.133910e-09  0.000003   \n",
       "1   MEME-1  OxyR_P.aeruginosa  EXPREG_00001560  1.568880e-07  0.000093   \n",
       "2   MEME-1             rpoD17           rpoD17  4.909700e-07  0.000290   \n",
       "3   MEME-1               RpoN         MX000100  1.036980e-06  0.000613   \n",
       "4   MEME-1               PhoP         MX000057  1.156850e-06  0.000684   \n",
       "5   MEME-1               OxyR         MX000154  1.986300e-06  0.001174   \n",
       "6   MEME-1             rpoD16           rpoD16  4.630600e-06  0.002737   \n",
       "7   MEME-1             rpoS17           rpoS17  5.842440e-06  0.003453   \n",
       "8   MEME-1               SigE         MX000068  7.194000e-06  0.004252   \n",
       "9   MEME-1             rpoD19           rpoD19  1.156250e-05  0.006833   \n",
       "10  MEME-2          ArcA_28-0        ArcA_28-0  6.043430e-08  0.000036   \n",
       "\n",
       "      qvalue  overlap  optimal_offset orientation             database  \n",
       "0   0.000006       37               9           +           dpinteract  \n",
       "1   0.000084       35               5           +             collectf  \n",
       "2   0.000132       25             -12           -           dpinteract  \n",
       "3   0.000207       24             -13           +             prodoric  \n",
       "4   0.000207       31              -3           +             prodoric  \n",
       "5   0.000266       31               3           -             prodoric  \n",
       "6   0.000496       19             -18           -           dpinteract  \n",
       "7   0.000569       29              -6           +           dpinteract  \n",
       "8   0.000643       26              11           -             prodoric  \n",
       "9   0.000885       22             -15           -           dpinteract  \n",
       "10  0.000069       26               2           -  SwissRegulon_e_coli  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.motif_info['SaeR-1'].matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, os.path.join('/home/jovyan/work/','processed_data','saureus.json.gz'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
